hw4/04-templates.js

```diff
@@ -1,64 +1,101 @@
+const formatCapital = (country) => {
+    return country.capital.length === 0
+        ? `${country.name.common} - no data`
+        : `${country.name.common} - ${country.capital}`
+}
 
-  res.render('page', {
-    heading: 'Countries and Capitals',
-    results: countries,
-  });
-});
+app.get('/capitals', (req, res) => {
+    // map the output array to create an array with country names and capitals
+    // check for empty data in the output array
+
+    axios.get(`${allUrl}?fields=name;capital`).then((response) => {
+        res.render('page', {
+            heading: 'Countries and Capitals',
+            results: response.data.sort(sortCountries).map(formatCapital),
+        })
+    })
+})
+
+const filterPopulation = (country) => {
+    return country.population >= 50000000
+}
+const sortPopulation = (first, second) => {
+    return second.population - first.population
+}
+
+const formatPopulation = (country) => {
+    return `${country.name.common} - ${country.population.toLocaleString()}`
+}
 
 app.get('/populous', (req, res) => {
-  // filter the output array for the countries with population of 50 million or more
-  // sort the resulting array to show the results in order of population
-  // map the resulting array into a new array with the country name and formatted population
-
-  let populous = ['China', 'India', 'United States of America'];
-
-  res.render('page', {
-    heading: 'Most Populous Countries',
-    results: populous,
-  });
-});
-
-app.get('/regions', (req, res) => {
-  // reduce the output array in a resulting object that will feature the numbers of countries in each region
-  // disregard empty data from the output array
-
-  let regions = ['Asia - 50', 'Europe - 53', 'Africa - 60'];
-
-  res.render('page', {
-    heading: 'Regions of the World',
-    results: regions,
-  });
-});
+    // filter the output array for the countries with population of 50 million or more
+    // sort the resulting array to show the results in order of population
+    // map the resulting array into a new array with the country name and formatted population
+    axios.get(`${allUrl}`).then((response) => {
+        res.render('page', {
+            heading: 'Most Populous Countries',
+            results: response.data
+                .filter(filterPopulation)
+                .sort(sortPopulation)
+                .map(formatPopulation),
+        })
+    })
+})
+
+const formatRegion = (regionData, name) => {
+    return `${name} - ${regionData.data.length}`
+}
+
+app.get('/regions', async (req, res) => {
+    // reduce the output array in a resulting object that will feature the numbers of countries in each region
+    // disregard empty data from the output array
+
+    const regions = ['Asia', 'Europe', 'Africa', 'Oceania', 'Americas']
+    const allRegions = []
+
+    Promise.all(
+        regions.map(async (region) => {
+            const regionData = await axios.get(`${regionUrl}/${region}`)
```

Comment:
my only concern is your code is not as efficient as it could be. sure it does the job, but it is making API call for almost the same data. making API call is costly. if you have the data, modify it in memory, instead of making API calls.

Reply:
And for this 84 line, you are making API call in the loop, which in general should be avoided.  No big deal since code is small.

Reply:
Props for using Promise.all